File: app/src/main/java/com/example/tochkatest/GoogleUseCase.java
package com.example.tochkatest;

import android.content.Intent;
import android.net.Uri;
import android.util.Log;
import android.widget.Toast;

import com.google.android.gms.auth.api.signin.GoogleSignIn;
import com.google.android.gms.auth.api.signin.GoogleSignInAccount;
import com.google.android.gms.auth.api.signin.GoogleSignInClient;
import com.google.android.gms.auth.api.signin.GoogleSignInOptions;
import com.google.android.gms.common.api.ApiException;
import com.google.android.gms.tasks.Task;

public class GoogleUseCase {


    private User user;
    final int RC_SIGN_IN = 0;
    private GoogleSignInClient mGoogleSignInClient;
    GoogleSignInOptions gso;
    private MainActivity mainActivity;
    private GoogleSignInAccount acct;

    public GoogleUseCase(MainActivity mainActivity) {

        this.mainActivity = mainActivity;
        gso = new GoogleSignInOptions.Builder(GoogleSignInOptions.DEFAULT_SIGN_IN)
                .requestEmail()
                .build();
        mGoogleSignInClient = GoogleSignIn.getClient(mainActivity, gso);


    }

    //google
    public void signIn(MainActivity mainActivity) {
        Intent signInIntent = mGoogleSignInClient.getSignInIntent();
        mainActivity.startActivityForResult(signInIntent, RC_SIGN_IN);
    }

    public void handleSignInResult(Task<GoogleSignInAccount> completedTask, MainActivity mainActivity ) {
        try {
            GoogleSignInAccount account = completedTask.getResult(ApiException.class);

            // Signed in successfully, show authenticated UI.
            acct = GoogleSignIn.getLastSignedInAccount(mainActivity);
            user = new User(acct.getGivenName() + " " + acct.getFamilyName(), acct.getPhotoUrl().toString());
            mainActivity.startActivity(new Intent(mainActivity, StartActivity.class).putExtra("USER", user));


        } catch (ApiException e) {
            // The ApiException status code indicates the detailed failure reason.
            // Please refer to the GoogleSignInStatusCodes class reference for more information.
            Log.e("TAG", "signInResult:failed code=" + e.getStatusCode());
        }
    }

    public void onActivityResult(int requestCode, int resultCode, Intent data){
        if (requestCode == RC_SIGN_IN) {
            // The Task returned from this call is always completed, no need to attach
            // a listener.
            Task<GoogleSignInAccount> task = GoogleSignIn.getSignedInAccountFromIntent(data);
            handleSignInResult(task,mainActivity);
        }
    }




}

File: app/src/main/java/com/example/tochkatest/useCases/GoogleUseCase.kt
package com.example.tochkatest.useCases

import android.content.Intent
import android.util.Log
import com.example.tochkatest.view.MainActivity
import com.example.tochkatest.view.StartActivity
import com.example.tochkatest.model.User

import com.google.android.gms.auth.api.signin.GoogleSignIn
import com.google.android.gms.auth.api.signin.GoogleSignInAccount
import com.google.android.gms.auth.api.signin.GoogleSignInClient
import com.google.android.gms.auth.api.signin.GoogleSignInOptions
import com.google.android.gms.common.api.ApiException
import com.google.android.gms.tasks.Task

class GoogleUseCase(private val mainActivity: MainActivity) {


    private var user: User? = null
    private val RC_SIGN_IN = 0
    private val mGoogleSignInClient: GoogleSignInClient
    private var gso: GoogleSignInOptions = GoogleSignInOptions.Builder(GoogleSignInOptions.DEFAULT_SIGN_IN)
            .requestEmail()
            .build()
    private var acct: GoogleSignInAccount? = null

    init {
        mGoogleSignInClient = GoogleSignIn.getClient(mainActivity, gso)


    }

    //google
    fun signIn(mainActivity: MainActivity) {
        val signInIntent = mGoogleSignInClient.signInIntent
        mainActivity.startActivityForResult(signInIntent, RC_SIGN_IN)
    }

    private fun handleSignInResult(completedTask: Task<GoogleSignInAccount>, mainActivity: MainActivity) {
        try {
            val account = completedTask.getResult(ApiException::class.java)

            // Signed in successfully, show authenticated UI.
            acct = GoogleSignIn.getLastSignedInAccount(mainActivity)
            user = User(acct!!.givenName + " " + acct!!.familyName, acct!!.photoUrl!!.toString())
            mainActivity.startActivity(Intent(mainActivity, StartActivity::class.java).putExtra("USER", user))


        } catch (e: ApiException) {
            // The ApiException status code indicates the detailed failure reason.
            // Please refer to the GoogleSignInStatusCodes class reference for more information.
            Log.e("TAG", "signInResult:failed code=" + e.statusCode)
        }

    }

    fun onActivityResult(requestCode: Int, resultCode: Int, data: Intent) {
        if (requestCode == RC_SIGN_IN) {
            // The Task returned from this call is always completed, no need to attach
            // a listener.
            val task = GoogleSignIn.getSignedInAccountFromIntent(data)
            this.handleSignInResult(task, mainActivity)
        }
    }


}
